email/fetcher.go

```diff
@@ -0,0 +1,393 @@
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		// skip messages without body
+		//
+		// TODO: side-effect from UidFetch and can probably be avoided
+		if !hasBody(msg) {
+			logger.Debugf("skip message due to not having a body (expected)")
+			err := f.persistSkipMessage(mailbox, msg)
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		toUnsee.AddNum(msg.Uid)
+		err := f.persistMessage(mailbox, msg, section)
+		if err != nil {
+			logger.Errorf("Failed to persist %v, error: %v\n", msg.Uid, err)
+		}
+	}
+
+	// unsee messages
+	flags := []interface{}{imap.SeenFlag}
+	err = email.UidStore(toUnsee, "-FLAGS.SILENT", flags, nil)
+	if err != nil && !strings.Contains(err.Error(), "Could not parse command") {
+		logger.Debugf("Failed to unsee messages, error: %v\n", err)
+	} else {
+		logger.Debugln("Successfully unseen messages")
+	}
+
+	return <-done
+}
+
+// persistMessage will persist the given message in the abuse scanner database
+func (f *Fetcher) persistMessage(mailbox *imap.MailboxStatus, msg *imap.Message, section *imap.BodySectionName) error {
+	// convenience variables
+	abuseDB := f.staticDatabase
+
+	// build the uid
+	uid := buildMessageUID(mailbox, msg.Uid)
+
+	// read the entire message body
+	bodyLit := msg.GetBody(section)
+	if bodyLit == nil {
+		return fmt.Errorf("msg %v has no body", uid)
+	}
+	body, err := ioutil.ReadAll(bodyLit)
+	if err != nil {
+		return errors.AddContext(err, "could not read msg body")
+	}
+
+	// parse the 'from'
+	from := "unknown"
+	if len(msg.Envelope.From) > 0 {
+		from = msg.Envelope.From[0].Address()
+	}
+
+	// create the email entity from the message
+	email := database.AbuseEmail{
+		ID:        primitive.NewObjectID(),
+		UID:       uid,
+		UIDRaw:    msg.Uid,
+		Body:      body,
+		From:      from,
+		Subject:   msg.Envelope.Subject,
+		MessageID: msg.Envelope.MessageId,
+
+		Parsed:    false,
+		Blocked:   false,
+		Finalized: false,
+
+		InsertedAt: time.Now(),
+	}
+
+	// insert the message in the database
+	err = abuseDB.InsertOne(email)
+	if err != nil {
+		return errors.AddContext(err, "could not insert email")
+	}
+	return nil
+}
+
+// persistSkipMessage will persist the given message as finalized in the abuse
+// scanner database, this ensures the message won't be considered 'missing'
+func (f *Fetcher) persistSkipMessage(mailbox *imap.MailboxStatus, msg *imap.Message) error {
+	// convenience variables
+	abuseDB := f.staticDatabase
+
+	// build the uid
+	uid := buildMessageUID(mailbox, msg.Uid)
+
+	// skip if it already exists
+	existing, err := abuseDB.FindOne(uid)
+	if err != nil {
+		return err
+	}
+	if existing != nil {
+		return nil
+	}
+
+	// create the email entity from the message
+	email := database.AbuseEmail{
+		ID:     primitive.NewObjectID(),
+		UID:    uid,
+		UIDRaw: msg.Uid,
+
+		Parsed:    true,
+		Blocked:   true,
```

Comment:
This should be an argument to the function. We don't block stuff based on emails we ourselves send.

I agree that this is not very important right now but we might want to build statistics on the number of links we block and this will artificially inflate that number.

Reply:
Can you elaborate? I don't understand the comment/suggestion here. How should it be an argument to the function? This won't stand in the way of statistics, I just need a way to persist which messages we want to skip. I would prefer to avoid adding a collection for that.

I agree it's not the cleanest but at the same time, I'm not sure whether I feel that strongly. I'll add a dedicated field to indicate that it's a `skip` message. That way it's even more easy to either clean it up in the future, or keep it and exclude all those from statistics with an easy `match`.

Reply:
I didn't express what I meant very well.

I meant that the `persistSkipMessage` method is marking a message as `finalized` but at the same time it's marking it as `blocked` which might not be the case. I think any message that reaches this code will already be `parsed` but some might be `skipped` while others will be `blocked`. The idea was to add an argument that allows the caller to set the `blocked` state, so we can easily aggregate on that in the future.